core/engine/src/value/inner/nan_boxed.rs

```diff
@@ -0,0 +1,864 @@
+//! This `[JsValue]` inner type is a NaN-boxed value, which is a 64-bits value
+//! that can represent any JavaScript value. If the integer is a non-NaN value,
+//! it will be stored as a 64-bits float. If it is a `f64::NAN` value, it will
+//! be stored as a quiet `NaN` value. Subnormal numbers are regular float.
+//!
+//! For any other type of values, the value will be stored as a 51-bits non-zero
+//! integer.
+//!
+//! In short, the memory layout of a NaN-boxed value is as follows:
+//!
+//! | Type of           | Bit Layout | Comment |
+//! |-------------------|------------|---------|
+//! | `+Infinity`       | `7FF0:0000:0000:0000`    | |
+//! | `-Infinity`       | `FFF0:0000:0000:0000`    | |
+//! | `NAN` (quiet)     | `7FF8:0000:0000:0000`    | |
+//! | `NAN` (signaling) | `FFF8:0000:0000:0000`    | |
+//! | `Undefined`       | `7FF4:0000:0000:0000`    | |
+//! | `Null`            | `7FF5:0000:0000:0000`    | |
+//! | `False`           | `7FF6:0000:0000:0000`    | |
+//! | `True`            | `7FF6:0000:0000:0001`    | |
+//! | `Integer32`       | `7FF7:0000:IIII:IIII`    | 32-bits integer. |
+//! | `BigInt`          | `7FF[8-F]:PPPP:PPPP:PPPP \| 0` | 51-bits pointer. Assumes non-null pointer. |
+//! | `Object`          | `7FF[8-F]:PPPP:PPPP:PPPP \| 1` | 51-bits pointer. |
+//! | `Symbol`          | `7FF[8-F]:PPPP:PPPP:PPPP \| 2` | 51-bits pointer. |
+//! | `String`          | `7FF[8-F]:PPPP:PPPP:PPPP \| 3` | 51-bits pointer. |
+//! | `Float64`         | Any other values.        | |
+//!
+//! Pointers have the highest bit (in the `NaN` tag) set to 1, so they
+//! can represent any value from `0x8000_0000_0000` to `0xFFFF_FFFF_FFFF`.
+//! The last 2 bits of the pointer is used to store the type of the value.
+//!
+//! The pointers are assumed to never be NULL, and as such no clash
+//! with regular NAN should happen.
+//!
+//! This only works on 4-bits aligned values, which is asserted when the
+//! `InnerValue` is created.
+
+use crate::{JsBigInt, JsObject, JsSymbol, JsVariant};
+use boa_gc::{custom_trace, Finalize, Trace};
+use boa_string::JsString;
+use core::fmt;
+use static_assertions::const_assert;
+
+/// Transform an `u64` into `f64`, by its bytes. This is necessary for
+/// keeping the MSRV at 1.82, as `f64::from_bits` is not const until
+/// 1.83.
+#[inline]
+const fn f64_from_bits(bits: u64) -> f64 {
+    unsafe { std::mem::transmute(bits) }
+}
```

Comment:
I would be in favor of just switching to 1.83 MSRV anyways. We always have been using mostly the latest version.

Reply:
We cannot, not until Tarpaulin is supporting it. See our own comment at https://github.com/boa-dev/boa/blob/main/.github/workflows/rust.yml#L35-L36

This PR is still opened: https://github.com/xd009642/tarpaulin/issues/1642

TBH I would be a bigger advocate to keep the MSRV constant. It's always a pain on dependents to move version. Let's make it a bigger deal when we update. Those methods aren't much to maintain. I can add a TODO to it (and create an issue) so we don't easily forget to remove them when MSRV moves past 1.83.